source/common/upstream/upstream_impl.cc

```diff
@@ -645,6 +645,128 @@ ClusterInfoImpl::ResourceManagers::load(const envoy::api::v2::Cluster& config,
       runtime, runtime_prefix, max_connections, max_pending_requests, max_requests, max_retries)};
 }
 
+PriorityStateManager::PriorityStateManager(ClusterImplBase& cluster,
+                                           const LocalInfo::LocalInfo& local_info,
+                                           const absl::optional<std::string> error_prefix)
+    : priority_state_(1), parent_(cluster), local_info_node_(local_info.node()) {
+  if (error_prefix.has_value()) {
+    Config::Utility::checkLocalInfo(error_prefix.value(), local_info);
+  }
+}
+
+void PriorityStateManager::initializePriorityFor(
+    const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint) {
+  const uint32_t priority = locality_lb_endpoint.priority();
+  if (priority_state_.size() <= priority) {
+    priority_state_.resize(priority + 1);
+  }
+  ASSERT(priority < priority_state_.size());
+  if (priority_state_[priority].first == nullptr) {
+    priority_state_[priority].first.reset(new HostVector());
+  }
+  if (locality_lb_endpoint.has_locality() && locality_lb_endpoint.has_load_balancing_weight()) {
+    // TODO(dio): Add tests as per @alyssawilk's comment here
+    // https://github.com/envoyproxy/envoy/pull/3261/files#r188292098
+    priority_state_[priority].second[locality_lb_endpoint.locality()] =
```

Comment:
Interestingly this looks like it's covered, but there's a few other spots missing.  It'd be great to have unit tests as part of this refactor.

Reply:
:+1: